linkerd/app/inbound/src/detect.rs

```diff
@@ -0,0 +1,323 @@
+        FSvc: svc::Service<io::BoxedIo, Response = ()> + Send + 'static,
+        FSvc::Error: Into<Error>,
+        FSvc::Future: Send,
+    {
+        self.map_stack(|cfg, rt, http| {
+            let detect_timeout = cfg.proxy.detect_protocol_timeout;
+
+            http.check_new_service::<Http, _>()
+                .push_map_target(|(http, tls)| Http { http, tls })
+                .push(svc::UnwrapOr::layer(
+                    // When HTTP detection fails, forward the connection to the application as
+                    // an opaque TCP stream.
+                    svc::stack(forward.clone()).into_inner(),
+                ))
+                .push_on_response(svc::MapTargetLayer::new(io::BoxedIo::new))
+                .push(svc::BoxNewService::layer())
+                .push_map_target(detect::allow_timeout)
+                .push(detect::NewDetectService::layer(cfg.proxy.detect_http()))
+                .check_new_service::<Tls, _>()
+                .push(rt.metrics.transport.layer_accept())
+                .check_new_service::<Tls, _>()
+                .push_request_filter(|(tls, t): (tls::ConditionalServerTls, T)| {
+                    match (t.param(), &tls) {
+                        // Permit all connections if no authentication is required.
+                        (AllowPolicy::Unauthenticated { .. }, _) => Ok(Tls::from_params(&t, tls)),
+                        // Permit connections with a validated client identity if authentication
+                        // is required.
+                        (
+                            AllowPolicy::Authenticated,
+                            tls::ConditionalServerTls::Some(tls::ServerTls::Established {
+                                client_id: Some(_),
+                                ..
+                            }),
+                        ) => Ok(Tls::from_params(&t, tls)),
+                        // Otherwise, reject the connection.
+                        _ => {
+                            let OrigDstAddr(a) = t.param();
+                            Err(IdentityRequired(a.port()))
+                        }
+                    }
+                })
+                .check_new_service::<(tls::ConditionalServerTls, T), _>()
+                .push(svc::BoxNewService::layer())
+                .push(tls::NewDetectTls::layer(TlsParams {
+                    timeout: tls::server::Timeout(detect_timeout),
+                    identity: rt.identity.clone(),
+                }))
+                .push_switch(
+                    // If this port's policy indicates that authentication is not required and
+                    // detection should be skipped, use the TCP stack directly.
+                    |t: T| -> Result<_, Infallible> {
+                        if let AllowPolicy::Unauthenticated { skip_detect: true } = t.param() {
+                            return Ok(svc::Either::B(t));
+                        }
```

Comment:
we now use the policy to determine when to skip detection (this is tested below).